bhchen2001-proposal-submission

bhchen2001 opened the pull request:
<!-- Please fill the following checkboxes -->
---
A good PR should include the following parts:

- [x] Provide a clear title
- [x] Provide meaningful descriptions in the PR and the commit logs
- [x] Name your branch like ``<github-user-name>-hw#``, e.g., ``myname-hw2-attempt1``
- [x] Merge from your specific branch (***not master***) to the `hw*` branch (***not master***, again)
- [x] Exclude irrelevant files, such as binaries, text files, or dot files
- [x] Make sure tests/CI pass

Commit:
doc: proposal initialization

Commit:
doc: fix for Github presentation

Commit:
doc: add entry in project list

yungyuc commented on project/README.rst:
```diff
@@ -42,3 +42,7 @@ Follow the format to add your project:
    `Project subject <github_account_name/README.rst>`__:
    https://github.com/github_account_name/project_name
+
+2. `bhchen2001 <https://github.com/bhchen2001>`__ for
```
Beautiful

yungyuc commented on project/bhchen2001/README.rst:
```diff
@@ -0,0 +1,95 @@
+
+1. A high-speed data structure for storing the vonoroi diagram and searching the nearby points.
+2. K-d tree for decomposing the points into different blocks.
+3. ``Fortune's algorithm`` for generating the vonoroi diagram independently at each thread on the blocks assigned to it.
```
This requires familiarity with the Fortune algorithm and multi-threading code development.

How familiar you are with the algorithm?

How familiar you are with a threading library?  Which one it is?

yungyuc commented on project/bhchen2001/README.rst:
```diff
@@ -0,0 +1,95 @@
+For the parallel version, the following requirements should be satisfied:
+
+1. A high-speed data structure for storing the vonoroi diagram and searching the nearby points.
+2. K-d tree for decomposing the points into different blocks.
```
Did you have experience with a spatial index?  Are you familiar with any existing source code implementing a k-d tree?

A k-d tree alone suffices a term project.

yungyuc commented on project/bhchen2001/README.rst:
```diff
@@ -0,0 +1,95 @@
+API Description
+===============
+
+The ``C++`` API::
```
You need a data structure to hold the Voronoi diagram, i.e., your `VoronoiDiagram` class.  Please briefly describe the data structure.

yungyuc commented on project/bhchen2001/README.rst:
```diff
@@ -0,0 +1,95 @@
+========================
+Parallel Vonoroi Diagram
```
It's Voronoi diagram, not "Vonoroi": https://mathworld.wolfram.com/VoronoiDiagram.html

yungyuc commented on project/bhchen2001/README.rst:
```diff
@@ -0,0 +1,95 @@
+Schedule
+========
+
+* Week 1 (04/08): Research on ``fortune's algorithm`` and ``k-d tree``.
```
I recommend you scale down the scope.  You are proposing to do two things: Fortune's algorithm for making a Voronoi diagram and k-d tree for indexing the geometry.  You are also proposing to parallelize the code.  While it is possible to prototype all of them in 8 weeks, it is very difficult to engineer it well.

If you insist to do everything, it's fine, but please find me at [sciwork #nsd channel](https://discord.gg/tZsUnx4XmB) to discuss before you make the decision.

yungyuc commented on project/bhchen2001/README.rst:
```diff
@@ -0,0 +1,95 @@
+  * Implement ``get_neighbors`` function.
+* Week 6 (05/13): Implement the parallel version.
+* Week 7 (05/20):
+  * Testing and debugging for the parallel version.
```
Putting testing at such a late stage tells me that you do not have experience to use testing for productivity.  Please discuss with me or the sciwork community for how testing can help you, and move it earlier in your plan.

yungyuc reviewed (changes requested):
- [x] Correct typo (e.g., Voronoi)
- [x] Clarify familiarity with k-d tree
- [x] Clarify familiarity wth Fortune's algorithm
- [ ] Clarify familiarity with threading
- [ ] Briefly describe the data structure for Voronoi diagram
- [ ] Reconsider testing in your plan

It baffles to name your branch exactly the same as the upstream branch (project), but it's too late for you to change. Please avoid it in the future.

When you are ready for the next review, leave a comment, clearly write down something like "I am ready, please review again".

yungyuc commented:
@bhchen2001 Please resolve the conflict and update the proposal.

Commit:
docs: update proposal
1. change topic from 'parallel voronoi' to 'search in kd tree'
2. schedule test/refactor earlier when developing

Commit:
docs: change proposal name, add github url

Commit:
docs: remove dash from topic

bhchen2001 committed:
Merge branch 'project' into project

bhchen2001 commented:
@yungyuc 
I scaled down my topic from 'parallel Voronoi diagram' to 'searching based on kd-tree' and advanced the testing process in the development phase. Please review it again and let me know where improvements are needed.

yungyuc commented on project/bhchen2001/README.rst:
```diff
@@ -0,0 +1,153 @@
+    // class to represent a point in multi-dimensional space
+    class Point{
+        public:
+            std::vector<double> coordinates;
```
`std::vector` requires dynamic memory allocation and it will slows down your system.  Be aware of this thing.

yungyuc commented on project/bhchen2001/README.rst:
```diff
@@ -0,0 +1,153 @@
+  * Implement ``Insert`` and ``Delete`` functions
+  * Design input data for ``KNN`` and ``RangeSearch`` functions
+* Week 4 (04/29)
+  * Testing and refactoring for ``Insert`` and ``Delete`` functions
```
Testing on the 4th week is not taking the most productivity to you, but not too bad.

yungyuc reviewed (approved):
This is OK
